docs/source/release_roadmap.org

```diff
@@ -0,0 +1,363 @@
+history into a single genesis block. However, it does not retain
+enough flexibility to be used as a single ledger for both private and
+public blockchains. And the following issues have been identified.
+
+1. The more efficient ISI-based language is not usable. The
+   expressions are not sufficiently expressive.
+
+   While this can be somewhat mitigated by the introduction of the
+   DSL, the main expressiveness-boosting facilities of ISI are not
+   usable, and thus only add overhead. A thorough exploration of a
+   restructuring of the Iroha Instruction Set Architecture is needed
+   and shall be published as an RFC in August 2023.
+
+2. WASM is a sub-optimal medium for storage of executable logic. It
+   alone is responsible for 80% of the complexity present in the code
+   base and around 95% of technical debt is related to some mitigation
+   of a shortcoming of WASM in the blockchain.
+
+   The advantages and disadvantages for various media storing Turing
+   complete logic on chain must also be discussed and shall be
+   published as an RFC by the end of August 2023.
+
+3. Iroha lacks any facilities for correlating the consensus process
+   with the blockchain data. This prevents Proof-of-work,
+   Proof-of-stake, slashing, governance and democracy from being
+   implemented on-chain.
+
+4. Iroha's consensus is largely inadequate for large networks. The
+   fault-tolerance and consensus speed scale poorly with network
+   size.
+
+5. An Iroha node has extremely high system requirements for
+   consumer-grade hardware. As of today the entire world state is
+   stored in RAM. The execution of transactions requires a
+   multi-threaded machine, but the throughput does not scale well
+   beyond 8 cores. This produces a set of system requirements that
+   are unlikely to occur naturally in a public blockchain.
+
+6. An Iroha network does not have a stable API. As such custom public
+   blockchains are liable to frequently squash their block history or
+   eschew upgrades.
+
+7. Iroha doesn't have a distinct understanding of the blockchain
+   uses. It caters to the private blockchains and as such has
+   validation for IP addresses, and some logic to handle mint-able and
+   non-mint-able assets as well as roles and permission tokens. This
+   information is hard-coded and non-modular, leading to public
+   blockchains missing vital functionality and having to commit their
+   changes upstream, and thus leading to incompatibility between the
+   Iroha APIs for a public and private blockchain.
+
+8. Iroha has no facilities to extend the expression language. This
+   dovetails into point 1, and leads to even less logic being encoded
+   as an ISI or =Expression= and instead relying heavily on WASM.
+
+9. WASM is only well-supported by Rust. While WASM is a
+   general-purpose binary format, only a few languages support it
+   natively: =C++=, =Go=, =nim=, =zig= and =Rust=. =Java=, =Kotlin=
```

Comment:
This languages cover pretty large share of used programming languages according to this [statistics](https://www.statista.com/statistics/793628/worldwide-developer-survey-most-used-languages/). Also i would noticed that this list is not [full](https://github.com/appcypher/awesome-wasm-langs).

Reply:
For one, this list already contains a half-truth: Rust only supports wasm32-unknown-unknown as a [tier 2 target](https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2) and wasm64 is a tier 3. 

Second, Java does not officially support it as a target, instead providing a "use-at-your-own-risk" workaround (you can see that in your second link), corroborated by [official WASM docs](https://webassembly.org/getting-started/developers-guide/). 

Thirdly, I don't think that that statistic supports your claim. In fact, the first item that technically qualifies (see second point) is Java, which ranks 7-th. So at best, we cover only 30% of developers worldwide, unless we start shipping v8 and allowing people to write financial code in JavaScript (not saying it couldn't be done, but it's like hammring a nail with a screwdriver). 

Fourth, awesome-* listicles are not reliable sources of information. One such listicle listed emacs as a terminal editor, gave Vim and Neovim one entry (despite the two projects being extremely different) and listed VSCode and VSCodium as two separate entries, despite the former being a proprietary version of the latter.

Reply:
I will admit that the phrasing makes it look like this list is meant to be exhaustive. I'll fix that.

Reply:
Unfortunately, most of stuff mentioned as 'supporting WASM' provide hardly usable, very limited, experimental prototypes. So it rather indicates whether some attempts were made in that direction at all, not more. Agree with Aleksandr on that.